ndt7/handler/handler.go

```diff
@@ -65,13 +70,41 @@ func (h Handler) downloadOrUpload(writer http.ResponseWriter, request *http.Requ
 	// and close(2), we'll end up keeping sockets that caused an error in the
 	// code above (e.g. because the handshake was not okay) alive for the time
 	// in which the corresponding *os.File is kept in cache.
-	defer warnonerror.Close(conn, "download: ignoring conn.Close result")
+	defer warnonerror.Close(conn, "downloadOrUpload: ignoring conn.Close result")
 	logging.Logger.Debug("downloadOrUpload: opening results file")
 	resultfp, err := results.OpenFor(request, conn, h.DataDir, kind)
 	if err != nil {
 		return // error already printed
 	}
-	defer warnonerror.Close(resultfp, "download: ignoring resultfp.Close result")
+
+	// Collect test metadata.
+	clientAddr := conn.RemoteAddr().(*net.TCPAddr)
+	serverAddr := conn.LocalAddr().(*net.TCPAddr)
+	result := &data.NDTResult{
+		GitShortCommit: prometheusx.GitShortCommit,
+		Version:        version.Version,
+		ClientIP:       clientAddr.IP.String(),
+		ClientPort:     clientAddr.Port,
+		ServerIP:       serverAddr.IP.String(),
+		ServerPort:     serverAddr.Port,
+		StartTime:      time.Now(),
+	}
+	// Guarantee that we record an end time, even if tester panics.
+	defer func() {
+		result.EndTime = time.Now()
```

Comment:
I am confused here because it seems you are setting `EndTime` already by `defer resultfp.EndTest` when you are in `download/measurer/measurer.go`. Also, I'd be happier if we could add more abstraction and not expose `Data` but just make it an implementation detail of `resultfp` accessible via APIs.

Reply:
Like, ideally, here I'd invoke a constructor of initialiser of `resultfp` and then perhaps `defer resultfp.WriteBack` or something similar, rather than having this long deferred function.